processor/src/main/java/org/mapstruct/ap/processor/MethodRetrievalProcessor.java

```diff
@@ -80,19 +93,60 @@ public int getPriority() {
         return 1;
     }
 
+    private List<SourceMethod> retrievePrototypeMethods(TypeMirror typeMirror, MapperConfig mapperConfig) {
+        if ( typeMirror == null || typeMirror.getKind() == TypeKind.VOID ) {
+            return Collections.emptyList();
+        }
+
+        TypeElement typeElement = asTypeElement( typeMirror );
+        List<SourceMethod> methods = new ArrayList<SourceMethod>();
+        for ( ExecutableElement executable : getAllEnclosedExecutableElements( elementUtils, typeElement ) ) {
+
+            ExecutableType methodType = typeFactory.getMethodType( typeElement, executable );
+            List<Parameter> parameters = typeFactory.getParameters( methodType, executable );
+            boolean containsTargetTypeParameter = SourceMethod.containsTargetTypeParameter( parameters );
+
+            // prototype methods don't have prototypes themselves
```

Comment:
> Would it be OK to discuss / change that within a separate issue?

Reading all of the above, I'm not sure yet that we have an issue (ok. perhaps a communication problem :smiley: ).

1. the current solution was actually made to inherit the configuration for an 'update' method from a 'create' method (just as you use as an example above).

2. The current solution (I just checked, see method `SourceMethod#isSimilar`) has an `equals` method on the last line for checking result types. So no inheritance over there.

3. What the new solution introduces is defining an inheritance relation for the result type. My original concern: I would not like to limit that to abstract result types only (but I think that Gunnar agrees with that).

4. I agree with Andreas that the `AUTO` mechanism should not apply to mapping methods within one `Mapper` but only to methods in the `MappingConfig` class. The example (update / create method) is a nice example why it would fail and lead to ambiguity if you did.

5. Wrt to the remark:

> Yes, ok. Although it's seems a very special scenario.

Is it? I would say that an interface is normal behavior is to hide the implementation. So only have a 'create' and 'update' for the root level element and hiding all the other branch and leaf mappings would actually make sense. What happens now is that I have to consider MapsStruct logic (the `Mapper` interface) as internal-implementation logic, wrap it in some kind of class and use another interface to that class to control the access to it and have something that I can actually manage (lifecycle wise). So, take for instance  OSGi why couldn't only the `Mapper` be  exported and we keep the generated implementation and the `MapperConfig` internal?